### PR TITLE
fix: add netstring encode/decode/stream library

### LINKED ISSUE #1807: add netstring+Stream utility
**What is the Problem Being Solved?**

The kernel-to-subprocess protocol uses [netstrings](https://en.wikipedia.org/wiki/Netstring) to frame each message. In #1797 I discovered that the library we were using for this fails to handle partial chunks in the stream decoder, which breaks the protocol any time the delivery message is larger than the OS kernel's pipe buffer (typically about 8kb, but it varies by OS, which is why the test passed on Linux but failed on macOS).

I haven't been able to find any correct libraries that offer a Stream interface, so I have to write my own.

**Description of the Design**

Four functions:

* `encode`: takes a Buffer, returns a Buffer
* `decode`: takes a Buffer, returns a list of Buffers (for the completed netstrings) plus a `leftover` Buffer
* `createWriter`: creates a `Transform` stream that takes Buffers (each of which is a single payload) and emits Buffers (to be concatenated out to a byte pipe)
* `createReader`: creates a `Transform` stream that takes Buffers (which are chunks that do not necessarily align to netstring boundaries, but when all are received and concatenated, will form one or more netstrings) and emits Buffers (one payload each)

**Security Considerations**

The usual memory-consumption attack exists (the sender just writes `99999..` forever, and the recipient keeps buffering the supposed length prefix forever in hopes of seeing the termination colon). We control both sides of the pipe, however, so I'm not going to try to address it.

**Test Plan**

Unit tests.

### PR DESCRIPTION
**Pull request by warner**
For use by the kernel-worker protocol.

refs #1797
closes #1807

This is not yet complete: I still need tests for the streaming interfaces, which I'm not entirely sure how to write.

@kriskowal could you eyeball the implementation and maybe see if I'm on the right track? Any ideas about how I should go about the test?

### CODE REVIEW

**Comment by kriskowal** [comment]
For TChannel, we used a bounded search test to suss out boundary conditions in the framing protocol. This should be easier. https://github.com/uber/tchannel-node/blob/master/test/streaming_bisect.js

File: packages/SwingSet/src/netstring.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+// input is a byte pipe, output is a sequence of Buffers
+export function createReader() {
+  let buffered = Buffer.from('');
+
+  function transform(chunk, encoding, callback) {
+    assert.equal(encoding, 'buffer');
+    buffered = Buffer.concat([buffered, chunk]);
+    let res = null;
+    try {
```
Comment (kriskowal) [inline packages/SwingSet/src/netstring.js:76]: Node.js deoptimizes any function contain try/catch. This can be addressed by layering any function that contains try/catch such that it’s isolated in a very small inner frame.

File: packages/SwingSet/test/test-netstring.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    t.deepEqual(encoded, expBuf);
+  }
+
+  eq('', '0:,');
+  eq('a', '1:a,');
+  eq('abc', '3:abc,');
+  const umlaut = 'ümlaut';
+  t.is(umlaut.length, 6);
+  const umlautBuffer = Buffer.from(umlaut, 'utf-8');
```
Comment (kriskowal) [inline packages/SwingSet/test/test-netstring.js]: I’ve been leaning harder on web API’s to make these kinds of libraries easier to port. To that end, `Uint8Array`, `TextDecoder`, and `TextEncoder` are handy, albeit less convenient than `Buffer.from`.

File: packages/SwingSet/test/test-netstring.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      console.log(`got : ${encoded}`);
+      console.log(`want: ${expBuf}`);
+    }
+    t.deepEqual(encoded, expBuf);
+  }
+
+  eq('', '0:,');
+  eq('a', '1:a,');
+  eq('abc', '3:abc,');
+  const umlaut = 'ümlaut';
```
Comment (kriskowal) [inline packages/SwingSet/test/test-netstring.js]: Could also test `👨‍👨‍👧‍👧` or something equally elaborate to not just test very high code points but also combining characters.

File: packages/SwingSet/src/netstring.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+export function createWriter() {
+  function transform(chunk, encoding, callback) {
+    assert.equal(encoding, 'buffer');
+    callback(encode(chunk));
+  }
+  return new Transform({ transform });
+}
+
+// Input is a Buffer containing zero or more netstrings and maybe some
+// leftover bytes. Output is zero or more decoded Buffers, one per netstring,
```
Comment (kriskowal) [inline packages/SwingSet/src/netstring.js:32]: Might be clearer if s/decoded buffers/strings.

File: packages/SwingSet/test/test-netstring.js (hunk 2 above)
Comment (kriskowal) [reply packages/SwingSet/test/test-netstring.js]: It would also increase portability to frame this API in terms of async iterators and relieve the dependence on Node.js-specific streams. Node.js streams do implement async iterators so bridging should not be hard.

File: packages/SwingSet/test/test-netstring.js (hunk 3 above)
Comment (warner) [reply packages/SwingSet/test/test-netstring.js]: added this, thanks

File: packages/SwingSet/src/netstring.js (hunk 4 above)
Comment (warner) [reply packages/SwingSet/src/netstring.js:32]: but I don't think they are `string`s, right? they're strictly `Buffer` objects, if I understand correctly

File: packages/SwingSet/src/netstring.js (hunk 1 above)
Comment (warner) [reply packages/SwingSet/src/netstring.js:76]: does that mean minimizing the number of variables closed over by the function containing the try/catch?

**Comment by warner** [comment]
Ok, I got the code all working, and expanded the tests (and also squashed, sorry). Ready for proper review now.

I'm uncertain about the Buffer/string distinction; I agree that using U8IntArray would be more general, but it sounds like it might be slightly harder to use, and our use case is somewhat narrow (two functions, both under our control, both in a Node.js -based start-compartment). I'll follow your judgement, do you think I should change it from what it looks like now?

File: packages/SwingSet/src/netstring.js (hunk 4 above)
Comment (kriskowal) [reply packages/SwingSet/src/netstring.js:32]: Ah, you are right. I was confused by the symmetry and “decoded”. Would “extracted” be a better term? Decode implied bytes in, string out to me.

File: packages/SwingSet/src/netstring.js (hunk 1 above)
Comment (kriskowal) [reply packages/SwingSet/src/netstring.js:76]: No, minimizing the closure isn’t necessary.

**Review by kriskowal — APPROVED** [review]

**Comment by kriskowal** [comment]
No further changes necessary for portability or performance in this first cut.
